File: 백준/Gold/2239. 스도쿠/스도쿠.cc
#include <iostream>
#include <cstring>
using namespace std;

int board[9][9];
bool row[9][9] = {false};
bool col[9][9] = {false};
bool grid[3][3][9] = {false};

bool solve(int i, int j) {
    if (i == 9) return true;
    if (j == 9) return solve(i + 1, 0);
    if (board[i][j] != 0) return solve(i, j + 1);
    
    int gi = i / 3, gj = j / 3;
    for (int num = 1; num <= 9; ++num) {
        int idx = num - 1;
        if (!row[i][idx] && !col[j][idx] && !grid[gi][gj][idx]) {
            row[i][idx] = col[j][idx] = grid[gi][gj][idx] = true;
            board[i][j] = num;
            if (solve(i, j + 1)) return true;
            board[i][j] = 0;
            row[i][idx] = col[j][idx] = grid[gi][gj][idx] = false;
        }
    }
    return false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    for (int i = 0; i < 9; ++i) {
        string s;
        cin >> s;
        for (int j = 0; j < 9; ++j) {
            board[i][j] = s[j] - '0';
            int num = board[i][j];
            if (num != 0) {
                int idx = num - 1;
                row[i][idx] = true;
                col[j][idx] = true;
                grid[i/3][j/3][idx] = true;
            }
        }
    }
    
    solve(0, 0);
    
    for (int i = 0; i < 9; ++i) {
        for (int j = 0; j < 9; ++j) {
            cout << board[i][j];
        }
        cout << '\n';
    }
    
    return 0;
}

File: 백준/Gold/2239. 스도쿠/스도쿠.py
import sys

sys.setrecursionlimit(10 ** 5)


def input(): return sys.stdin.readline().rstrip()


board = [list(map(int, input())) for _ in range(9)]

row = [0] * 9
col = [0] * 9
grid = [0] * 9

empty = []

for r in range(9):
    for c in range(9):
        if board[r][c] != 0:
            val = 1 << (board[r][c] - 1)
            row[r] |= val
            col[c] |= val
            grid[r // 3 * 3 + c // 3] |= val
        else:
            empty.append((r, c))


def check(mask):
    full_mask = 0b111111111
    return full_mask & ~mask


def solution(depth):
    if depth == len(empty):
        for r in range(9):
            print(*board[r], sep='')
        exit(0)

    r, c = empty[depth]
    candidates = check(
        row[r] | col[c] | grid[r // 3 * 3 + c // 3])

    for num in range(1, 10):
        if candidates & (1 << (num - 1)):
            board[r][c] = num
            val = 1 << (num - 1)
            row[r] |= val
            col[c] |= val
            grid[r // 3 * 3 + c // 3] |= val

            solution(depth + 1)

            board[r][c] = 0
            row[r] &= ~val
            col[c] &= ~val
            grid[r // 3 * 3 + c // 3] &= ~val


solution(0)